model/map/src/main/java/org/keycloak/models/map/common/UuidValidator.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.models.map.common;
+
+import java.util.regex.Pattern;
+
+import javax.ws.rs.BadRequestException;
+
+import org.jboss.logging.Logger;
+import org.keycloak.Config;
+import org.keycloak.models.map.storage.MapStorageSpi;
+
+/**
+ * Utility class for validating and converting UUIDs.
+ *
+ * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
+ */
+public class UuidValidator {
+
+    protected static final Logger logger = Logger.getLogger(UuidValidator.class);
+
+    protected static final Pattern UUID_REGEX_PATTERN = Pattern.compile("^[0-9a-fA-F]{8}-([0-9a-fA-F]{4}-){3}[0-9a-fA-F]{12}$");
+
+    private UuidValidator() {}
+
+    /**
+     * Validates that the specified {@code id} is a {@code UUID}.
+     *
+     * @param id the {@code id} to be validated.
+     * @return {@code true} if the {@code id} is a {@code UUID}; {@code false} otherwise.
+     */
+    public static boolean validate(final String id) {
+        return (id == null) ? false : UUID_REGEX_PATTERN.matcher(id).matches();
+    }
+
+    /**
+     * Validates that the specified {@code id} is a {@code UUID}. If it is, the {@code id} itself is returned. If the {@code id}
+     * is {@code null}, a new {@code UUID} is generated and returned. If not, then the {@code mapStorage} configuration is checked
+     * to determine whether the specified {@code id} should be rejected or if a new {@code UUID} should be created and returned
+     * instead.
+     *
+     * @param id the {@code id} to be validated.
+     * @param component the name of the component (e.g. realm, component) whose id is being validated.
+     * @return the {@code id} itself if it is a valid {@code UUID}, or a new generated {@code UUID}.
+     * @throws {@code InvalidUuidException} if the {@code id} is invalid and the {@code mapStorage} has been configured to
+     * reject invalid ids.
+     */
+    public static String validateAndConvert(final String id, final String component) {
+        boolean rejectInvalid = Boolean.parseBoolean(Config.scope(MapStorageSpi.NAME).get("rejectInvalidIds"));
+        return validateAndConvert(id, component, rejectInvalid);
+    }
```

Comment:
Just thinking a bit more about this validation/conversion of ids into UUIDs for the JPA storage.. using a config property from the `mapStorage` does feel a bit misplaced as other storage impls might not have this hard requirement of ids being UUIDs.

Perhaps we should have this `rejectInvalidIds` config in the `jpa-map-storage` instead and inject its value into the created transactions. The transaction's `create` method can then run the `validateAndConvert` method using the flag that signals if an exception should be thrown or if the invalid id should be simply converted into an UUID.

Thoughts?

Reply:
If the UUIDs are a requirement for the map storage in general, then maybe the approach of using a config in `mapStorage` directly is not displaced at all.

Reply:
Given the example of LDAP, the UUIDs are no requirement across all map storages. A store might decide to create IDs differently, and as long as they are a String they are fine to do so. Having a UUID might be default for Keycloak, and the only supported way for JPA, but this won't hold true for all map storages.

Reply:
+1 to what @ahus1 wrote.

Also: can `rejectInvalidIds` ever be false? Wouldn't passing in an invalid ID result in an exception from the store later upon saving the object?

Reply:
@ahus1 @hmlnarik  Ok, so based on the discussion the conclusion is that the store has the freedom to convert ids at its own discretion, treating any supplied id as a suggestion. In that case, I should remove the id validation/conversion from the `MapRealmProvider` and `MapRealmAdapter` and move it to the JPA transaction `create` method. The reason being that converting at the logical layer is meaningless if other store, like LDAP, will have to convert it again.

In JPA, once the store is invoked to create an entity it can validate if the incoming id is a UUID and convert if needed. I'll also remove the exception and log message. Can I proceed this way?

Reply:
+1

Reply:
Done! All changes pushed based on the reviews

Reply:
This approach worked, got a clean testsuite run using this approach instead of adding things in the logical layer.